ext/oboe_metal/src/profiling.cc

```diff
@@ -0,0 +1,405 @@
+                                  empty,                           // <vector> new frames
+                                  0,                               // number of exited frames
+                                  prof_data_map[tid].prev_num,     // total number of frames
+                                  prof_data_map[tid].omitted,      // array of timestamps of omitted snapshots
+                                  prof_data_map[tid].omitted_num,  // number of omitted snapshots
+                                  tid);                            // thread id
+
+    prof_data_map[tid].omitted_num = 0;
+}
+
+void Profiling::process_snapshot(VALUE *frames_buffer, int num, pid_t tid, long ts) {
+    int num_new = 0;
+    int num_exited = 0;
+    vector<FrameData> new_frames;
+
+    num = Frames::remove_garbage(frames_buffer, num);
+
+    // find the number of matching frames from the top
+    int num_match = Frames::num_matching(frames_buffer,
+                                         num,
+                                         prof_data_map[tid].prev_frames_buffer,
+                                         prof_data_map[tid].prev_num);
+    num_new = num - num_match;
+
+    num_exited = prof_data_map[tid].prev_num - num_match;
+
+    if (num_new == 0 && num_exited == 0) {
+        prof_data_map[tid].omitted[prof_data_map[tid].omitted_num] = ts;
+        prof_data_map[tid].omitted_num++;
+
+        // the omitted buffer can fill up if the interval is small
+        // and the stack doesn't change
+        // We need to send a profiling event with the timestamps when it is full
+        if (prof_data_map[tid].omitted_num >= BUF_SIZE) {
+            Profiling::send_omitted(tid, ts);
+        }
+        return;
+    }
+
+    Frames::collect_frame_data(frames_buffer, num_new, new_frames);
+
+    Logging::log_profile_snapshot(prof_data_map[tid].md,
+                                  prof_data_map[tid].prof_op_id,
+                                  ts,                              // timestamp
+                                  new_frames,                      // <vector> new frames
+                                  num_exited,                      // number of exited frames
+                                  num,                             // total number of frames
+                                  prof_data_map[tid].omitted,      // array of timestamps of omitted snapshots
+                                  prof_data_map[tid].omitted_num,  // number of omitted snapshots
+                                  tid);                            // thread id
+
+    prof_data_map[tid].omitted_num = 0;
+    prof_data_map[tid].prev_num = num;
+    for (int i = 0; i < num; ++i)
+        prof_data_map[tid].prev_frames_buffer[i] = frames_buffer[i];
+}
+
+void Profiling::profiler_job_handler(void *data) {
+    static std::atomic_int in_job_handler;
+
+    if (in_job_handler) return;
+    in_job_handler++;
```

Comment:
right ... originally these were int in the single threaded version to avoid processing if intervals overlap, but with multiple threads it needs to be refined ... sounds like a case for a semaphore ... let me see

Reply:
bummer ... semaphores only in C++20